### PR TITLE
DBPW - Update PostgreSQL to adhere to v5 Database interface

### PR DESCRIPTION
**Pull request by pcman312**
# Overview
Updates the PostgreSQL Database plugin to adhere to the v5 Database interface.

### CODE REVIEW

File: plugins/database/postgresql/postgresql_test.go [hunk 1]
```diff
@@ -151,44 +225,29 @@ func TestPostgreSQL_CreateUser(t *testing.T) {
...
-			if err = testCredsExist(t, db.ConnectionURL, username, password); err != nil {
-				t.Fatalf("Could not connect with new credentials: %s", err)
-			}
+			test.credsAssertion(t, db.ConnectionURL, resp.Username, test.req.Password)
 		})
 	}
 }
 
-func TestPostgreSQL_RenewUser(t *testing.T) {
+func TestUpdateUser_expiration(t *testing.T) {
```
Comment (tvoran) [inline plugins/database/postgresql/postgresql_test.go:250]: Why are there two expiration tests? This one and `TestUpdateUser_Expiration()` on line 409.

**Review by tvoran — APPROVED** [review]

File: plugins/database/postgresql/postgresql.go [hunk 2]
```diff
@@ -104,54 +107,56 @@ func (p *PostgreSQL) getConnection(ctx context.Context) (*sql.DB, error) {
...
+		return newdbplugin.UpdateUserResponse{}, fmt.Errorf("no changes requested")
 	}
 
-	username = staticUser.Username
-	password = staticUser.Password
-	if username == "" || password == "" {
-		return "", "", errors.New("must provide both username and password")
+	merr := &multierror.Error{}
+	if req.Password != nil {
+		merr = multierror.Append(merr, p.changeUserPassword(ctx, req.Username, req.Password))
```
Comment (calvn) [inline plugins/database/postgresql/postgresql.go:120]: Nit: This is a bit more verbose, but having the call to `p.changeUserPassword` outside of `multierror.Append` makes it a bit easier read.

```suggestion
		err := p.changeUserPassword(ctx, req.Username, req.Password)
		merr = multierror.Append(merr, err)
```

File: plugins/database/postgresql/postgresql_test.go [hunk 3]
```diff
@@ -344,155 +431,234 @@ func TestPostgreSQL_RevokeUser(t *testing.T) {
...
+	}
+}
+
+func waitUntilCredsDoNotExist(timeout time.Duration) credsAssertion {
+	return func(t testing.TB, connURL, username, password string) {
+		t.Helper()
+		ctx, cancel := context.WithTimeout(context.Background(), timeout)
+		defer cancel()
+
+		ticker := time.NewTicker(10 * time.Millisecond)
```
Comment (calvn) [inline plugins/database/postgresql/postgresql_test.go:641]:
```suggestion
		ticker := time.NewTicker(10 * time.Millisecond)
		defer ticker.Stop()
```

File: plugins/database/postgresql/postgresql_test.go (hunk 1 above)
Comment (calvn) [reply plugins/database/postgresql/postgresql_test.go:250]: I'm also curious on this one

**Review by calvn — APPROVED** [review]
Looks good aside from Theron's question around one of the tests.

File: plugins/database/postgresql/postgresql.go (hunk 2 above)
Comment (pcman312) [reply plugins/database/postgresql/postgresql.go:120]: Good call. Done

File: plugins/database/postgresql/postgresql_test.go (hunk 3 above)
Comment (pcman312) [reply plugins/database/postgresql/postgresql_test.go:603]: Ah good catch. Fixed

File: plugins/database/postgresql/postgresql_test.go [hunk 4]
```diff
@@ -213,128 +283,251 @@ func TestPostgreSQL_RenewUser(t *testing.T) {
...
+			if actualExpiration.IsZero() {
+				t.Fatalf("Initial expiration is zero but should be set")
+			}
+			if !actualExpiration.Equal(initialExpiration) {
+				t.Fatalf("Actual expiration: %s Expected expiration: %s", actualExpiration, initialExpiration)
 			}
 
-			if !connProducer.Initialized {
-				t.Fatal("Database should be initialized")
+			newExpiration := test.newExpiration.Truncate(time.Second)
```
Comment (calvn) [inline plugins/database/postgresql/postgresql_test.go:420]: Should we truncate this value to the second so that we don't fail due to precision?

**Review by tvoran — APPROVED** [review]

**Pull request merged**